### PR TITLE
Index 4337 UserOperations

### LINKED ISSUE #1858: 4337 - Index User Operations
# Index UserOperation
## Database Model
UserOperation
- bytes32 userOpHash
- address sender
- address paymaster
- uint256 nonce
- bool success
- uint256 actualGasCost
- uint256 actualGasUsed
## Event 
`UserOperationEvent (index_topic_1 bytes32 userOpHash, index_topic_2 address sender, index_topic_3 address paymaster, uint256 nonce, bool success, uint256 actualGasCost, uint256 actualGasUsed)`

### LINKED ISSUE #1859: 4337 - Index Safe Operations
# Index Safe Operation
## Database Model
Safe Operation: https://github.com/safe-global/safe-modules/blob/main/modules/4337/contracts/Safe4337Module.sol#L55
## How to get the data
Depends on https://github.com/safe-global/safe-eth-py/issues/774

### LINKED ISSUE #1860: 4337 - Add new endpoints to return indexed SafeOperations
# Related issues
- #1859 

# Endpoint

## URL
`GET /api/v1/safes/<address>/safe-operations/`
`GET /api/v1/safes/safe-operations/<safeOpStructHash>/`

## Response
TBD
```json
{
 
}
```

### LINKED ISSUE #1861: 4337 - Post endpoint to gather signatures for SafeOperations
# Related issues
- #1859 

# Endpoint

## URL
`POST /api/v1/safes/<address>/safe-operations/`

## Payload
TBD

## Response
`201 CREATED`

### PR DESCRIPTION
**Pull request by Uxio0**
- Closes #1858
- Closes #1859 
- Closes #1860 
- Closes #1861

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -2,15 +2,15 @@
 # See https://pre-commit.com/hooks.html for more hooks
 repos:
   - repo: https://github.com/PyCQA/isort
-    rev: 5.12.0
+    rev: 5.13.2
```
Comment (moisses89) [inline .pre-commit-config.yaml:5]: In my opinion this should be updated in different PR because is adding too much changes that are not related with the focus of the PR.

Comment (Uxio0) [reply .pre-commit-config.yaml:5]: Sorry, I was having an error with the current version

Comment (Uxio0) [reply .pre-commit-config.yaml]: Pre-commit was already updated but I didn't rebased

File: safe_transaction_service/account_abstraction/tests/test_models.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+from django.test import TestCase
+
+
+class TestModels(TestCase):
+    pass
```
Comment (moisses89) [inline safe_transaction_service/account_abstraction/tests/test_models.py:5]: We are not including too much tests, I would agree if you create an issue to do it later in other PR.

File: safe_transaction_service/account_abstraction/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+from gnosis.safe.tests.safe_test_case import SafeTestCaseMixin
+
+from .. import models
+from . import factories
+
+logger = logging.getLogger(__name__)
+
+
+# FIXME Refactor datetime_to_str
+def datetime_to_str(value: datetime.datetime) -> str:
```
Comment (moisses89) [inline safe_transaction_service/account_abstraction/tests/test_views.py:35]: Maybe refactor in utils?

File: safe_transaction_service/account_abstraction/views.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    @swagger_auto_schema(
+        request_body=serializers.SafeOperationSerializer,
+        responses={201: "Created"},
+    )
+    def post(self, request, address, *args, **kwargs):
+        """
+        Create a new signed message for a Safe. Message can be:
+        - A ``string``, so ``EIP191`` will be used to get the hash.
+        - An ``EIP712`` ``object``.
```
Comment (moisses89) [inline safe_transaction_service/account_abstraction/views.py:64]: Looks like a copy paste documentation

**Review by moisses89 — CHANGES_REQUESTED** [review]
Good work :) :rocket:

File: safe_transaction_service/account_abstraction/serializers.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+from safe_transaction_service.utils.ethereum import get_chain_id
+
+from ..utils.serializers import get_safe_owners
+from .models import SafeOperation
+from .models import SafeOperation as SafeOperationModel
+from .models import SafeOperationConfirmation
+from .models import UserOperation as UserOperationModel
+
+SIGNATURE_LENGTH = 5_000
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/serializers.py:26]: I don't think configuring in settings is possible, as they need to be known during service startup, but definitely we should have the constant in just one place

File: safe_transaction_service/utils/ethereum.py [hunk 6]
```diff
@@ -1,9 +1,6 @@
-from functools import cache
-
 from gnosis.eth import EthereumClientProvider, EthereumNetwork
 
 
-@cache
 def get_chain_id() -> int:
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/utils/ethereum.py:4]: `EthereumClient` already caches the chain_id, no need to double caching

[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/utils/ethereum.py:4]: `EthereumClient` already caches the chain_id, no need to double caching

File: requirements.txt [hunk 7]
```diff
@@ -33,5 +33,5 @@ psycogreen==1.0.2
 psycopg2==2.9.9
 redis==5.0.3
 requests==2.31.0
-safe-eth-py[django]==6.0.0b22
+safe-eth-py[django] @ git+https://github.com/safe-global/safe-eth-py
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply requirements.txt:36]: Yes, I will update it before merging

File: safe_transaction_service/account_abstraction/services/aa_processor_service.py [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+        :param safe_address: Sender to check in UserOperation
+        :param ethereum_tx: EthereumTx to check for UserOperations
+        :return: Number of detected ``UserOperations`` in transaction
+        """
+        aa_logs = self.get_user_operation_logs(safe_address, ethereum_tx.logs)
+        number_detected_user_operations = len(aa_logs)
+        if not self.bundler_client:
+            logger.debug(
+                "Detected 4337 User Operation but bundler client was not configured"
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/services/aa_processor_service.py:344]: That's right

File: safe_transaction_service/account_abstraction/serializers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/serializers.py]: Done

File: safe_transaction_service/account_abstraction/views.py (hunk 4 above)
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/views.py:64]: It is 😂

**Review by falvaradorodriguez — COMMENTED** [review]
Great job! 👌

File: config/settings/base.py [hunk 9]
```diff
@@ -400,6 +401,9 @@
         "safe_transaction_service.history.services.balance_service": {
             "level": "DEBUG" if DEBUG else "WARNING",
         },
+        "safe_transaction_service.history.indexers.tx_processor": {
+            "level": "DEBUG",
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply config/settings/base.py:405]: Nice catch, I was using it to debug

File: .github/workflows/python.yml [hunk 10]
```diff
@@ -93,6 +93,7 @@ jobs:
         ETHEREUM_MAINNET_NODE: ${{ secrets.ETHEREUM_MAINNET_NODE }}
         ETHEREUM_NODE_URL: http://localhost:8545
         ETHEREUM_TRACING_NODE_URL: http://localhost:8545
+        ETHEREUM_4337_BUNDLER_URL: ${{ secrets.ETHEREUM_4337_BUNDLER_URL }}
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply .github/workflows/python.yml:96]: No, because I don't want people to be able to see the bundler url (as it includes our API key)

File: safe_transaction_service/account_abstraction/tests/test_models.py (hunk 2 above)
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/tests/test_models.py]: Done, included more tests

File: safe_transaction_service/account_abstraction/tests/test_views.py (hunk 3 above)
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/tests/test_views.py]: Done

File: config/settings/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply config/settings/base.py:405]: Fixed

File: safe_transaction_service/account_abstraction/serializers.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+
+# ================================================ #
+#            Request Serializers
+# ================================================ #
+class SafeOperationSerializer(serializers.Serializer):
+    nonce = serializers.IntegerField(min_value=0)
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/serializers.py:32]: You are right

File: safe_transaction_service/account_abstraction/models.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    call_data_gas_limit = Uint256Field()
+    verification_gas_limit = Uint256Field()
+    pre_verification_gas = Uint256Field()
+    max_fee_per_gas = Uint256Field()
+    max_priority_fee_per_gas = Uint256Field()
+    paymaster = EthereumAddressV2Field(
+        db_index=True, null=True, blank=True, editable=True
+    )
+    paymaster_data = models.BinaryField(null=True, blank=True, editable=True)
+    signature = models.BinaryField(null=True, blank=True, editable=True)
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/models.py:52]: I use BinaryField as I want the bytes output and not a HexStr

File: safe_transaction_service/account_abstraction/serializers.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/serializers.py:32]: @moisses89 This is a serializer, not a model. Integer fields works without an issue

File: safe_transaction_service/account_abstraction/serializers.py [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+        safe_address: ChecksumAddress,
+        safe_operation_hash: bytes,
+        safe_operation_hash_preimage: bytes,
+        signature: bytes,
+    ) -> List[SafeSignature]:
+        safe_owners = get_safe_owners(safe_address)
+        parsed_signatures = SafeSignature.parse_signature(
+            signature, safe_operation_hash, safe_operation_hash_preimage
+        )
+        signature_owners = set()
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/serializers.py:62]: Good suggestion

File: safe_transaction_service/account_abstraction/serializers.py [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+
+
+class SafeOperationResponseSerializer(serializers.Serializer):
+    created = serializers.DateTimeField(source="safe_operation.created")
+    modified = serializers.DateTimeField(source="safe_operation.modified")
+
+    sender = eth_serializers.EthereumAddressField()
+    user_operation_hash = eth_serializers.HexadecimalField(source="hash")
+    safe_operation_hash = eth_serializers.HexadecimalField(source="safe_operation.hash")
+    nonce = serializers.IntegerField(min_value=0)
```
[the remark this answers is not in the document]
Comment (Uxio0) [reply safe_transaction_service/account_abstraction/serializers.py:245]: This is a serializer, Integer works

**Review by moisses89 — APPROVED** [review]
